doc/services/input/gpio-kbd.rst

```diff
@@ -0,0 +1,240 @@
+                    <&gpio0 4 GPIO_ACTIVE_HIGH>,
+                    <&gpio0 5 GPIO_ACTIVE_HIGH>;
+        col-drive-inactive;
+        settle-time-us = <0>;
+        no-ghostkey-check;
+   };
+
+GPIO with no interrupt support
+******************************
+
+Some GPIO controllers have limitations on GPIO interrupts, and may not support
+enabling interrupts on all row GPIOs at the same time.
+
+In this case, the driver can be configured to not use interrupt at all, and
+instead idle by selecting all columns and keep polling on the row GPIOs, which
+is a single GPIO API operation if the pins are sequential.
+
+This configuration can be enabled by setting the ``idle-mode`` property to
+``poll``:
+
+.. code-block:: devicetree
+
+   kbd-matrix {
+        compatible = "gpio-kbd-matrix";
+        ...
+        idle-mode = "poll";
+   };
+
+GPIO multiplexer
+****************
+
+In more extreme cases, such as if the columns are using a multiplexer and it's
+impossible to select all of them at the same time, the driver can be configured
+to scan continuously.
+
+This can be done by setting ``idle-mode`` to ``scan`` and ``poll-timeout-ms``
+to ``0``.
+
+.. code-block:: devicetree
+
+   kbd-matrix {
+        compatible = "gpio-kbd-matrix";
+        ...
+        poll-timeout-ms = <0>;
+        idle-mode = "scan";
+   };
+
+Row and column GPIO selection
+*****************************
+
+If the row GPIOs are sequential and on the same gpio controller, the driver
+automatically switches API to read from the whole GPIO port rather than the
+individual pins. This is particularly useful if the GPIOs are not memory
+mapped, for example on an I2C or SPI port expander, as this significantly
+reduces the number of transactions on the corresponding bus.
+
+The same is true for column GPIOs, but only if the matrix is configured for
+``col-drive-inactive``, so that is only usable for matrixes with isolation
+diodes.
+
+16 bit row support
+******************
+
+The driver uses an 8 bit datatype to store the row state by default, which
+limits the matrix row size to 8. This can be increased to 16 by enabling the
+:kconfig:option:`CONFIG_INPUT_KBD_MATRIX_16_BIT_ROW` option.
+
+Actual key mask configuration
+*****************************
+
+If the key matrix is not complete, a map of the keys that are actually
+populated can be specified using the `actual-key-mask` property. This allows
+the matrix state to be filtered to remove keys that are not present before
+ghosting detection, potentially allowing key combinations that would otherwise
+be blocked by it.
+
+For example for a 3x3 matrix missing a key:
+
+.. figure:: no-sw4.svg
+      :align: center
+      :width: 50%
+
+      A 3x3 matrix missing a key.
+
+.. code-block:: devicetree
+
+   kbd-matrix {
+        compatible = "gpio-kbd-matrix";
+        ...
+        actual-key-mask = <0x07 0x05 0x07>;
+   };
+
+Would allow, for example, to detect pressing ``Sw0``, ``SW1`` and  ``SW3`` at
```

Comment:
```suggestion
This would allow, for example, to detect pressing ``Sw0``, ``SW1`` and  ``SW3`` at
```

Reply:
also, `SW0` does not exist (maybe you used different numbering scheme before), and I am terrible at ghosting so can't suggest the proper change :)

Reply:
Ah! yeah I originally had the illustration refdes numbered from 0, then I moved them around in the CAD and the thing automatically renumbered them (it tries REALLY hard to keep them unique), so I had to rename it again and I guess at that point I had my wannabe-EE hat on an started counting from 1, which is the right thing to do in this context I think, I'll rename them to 1, 2, 4.

I actually wanted to do an animation of the whole scanning story, looked a bit for online tools that could export something embeddable here but did not find any, too bad :-), the ghosting concept is very simple, I may add some extra illustration down the road.